File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/AverageHash.swift
import Foundation

/// ahash: convert to grayscale, Lanczos resize to NxN, threshold against mean.
public func averageHash(_ image: RGBImage, hashSize: Int) throws -> Hash {
    guard hashSize >= 2 else { throw ImageHashError.invalidHashSize(hashSize) }
    let rgb = ImgRGB.toRGB(image)
    let gray = rgbToGray(rgb.data, width: rgb.width, height: rgb.height)
    let resized = lanczosResize(gray, srcW: rgb.width, srcH: rgb.height, dstW: hashSize, dstH: hashSize)

    var sum = 0
    for v in resized { sum += Int(v) }
    let avg = Double(sum) / Double(hashSize * hashSize)

    var bits: [[Bool]] = []
    for y in 0..<hashSize {
        var row = [Bool](repeating: false, count: hashSize)
        for x in 0..<hashSize {
            row[x] = Double(resized[y * hashSize + x]) > avg
        }
        bits.append(row)
    }
    return try Hash(bits: bits)
}

/// Shared helper: convert a flat RGB [UInt8] (row-major triples) to a flat grayscale [UInt8].
/// Used by averageHash, dhash, phash, whashHaar (colorhash uses toGray directly per-pixel).
internal func rgbToGray(_ rgb: [UInt8], width: Int, height: Int) -> [UInt8] {
    var out = [UInt8](repeating: 0, count: width * height)
    var si = 0
    for i in 0..<(width * height) {
        out[i] = UInt8(toGray(Int(rgb[si]), Int(rgb[si + 1]), Int(rgb[si + 2])))
        si += 3
    }
    return out
}

File: hash/js/rosetta-squint-hash/src/averageHash.ts
import { Hash, ImageHashError, type RgbImage } from "./hash.js";
import { toRgb, validateRgbImage } from "./internal/imgRgb.js";
import { resize } from "./internal/lanczos.js";
import { toGray } from "./internal/pilGray.js";

/**
 * ahash: convert to grayscale, Lanczos resize to NxN, threshold against mean.
 */
export function averageHash(img: RgbImage, hashSize: number): Hash {
  validateRgbImage(img);
  if (!Number.isInteger(hashSize)) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be an integer, got ${hashSize}`);
  }
  if (hashSize < 2) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be >= 2, got ${hashSize}`);
  }
  const rgb = toRgb(img);
  const gray = rgbToGray(rgb.data, rgb.width, rgb.height);
  const resized = resize(gray, rgb.width, rgb.height, hashSize, hashSize);

  let sum = 0;
  for (let i = 0; i < resized.length; i++) sum += resized[i];
  const avg = sum / (hashSize * hashSize);

  const bits: boolean[][] = [];
  for (let y = 0; y < hashSize; y++) {
    const row: boolean[] = new Array(hashSize);
    for (let x = 0; x < hashSize; x++) {
      row[x] = resized[y * hashSize + x] > avg;
    }
    bits.push(row);
  }
  return new Hash(bits);
}

/** Shared helper: convert a flat RGB Uint8Array (row-major triples) to a flat grayscale Uint8Array. */
export function rgbToGray(rgb: Uint8Array, width: number, height: number): Uint8Array {
  const out = new Uint8Array(width * height);
  let si = 0;
  for (let i = 0; i < width * height; i++) {
    out[i] = toGray(rgb[si], rgb[si + 1], rgb[si + 2]);
    si += 3;
  }
  return out;
}
